mm2src/coins/utxo.rs

```diff
@@ -256,1602 +254,151 @@ pub struct UtxoCoinImpl {  // pImpl idiom.
-        Builder::build_p2pkh(&to.hash).to_bytes()
-    } else if to.prefix == coin.p2sh_addr_prefix && to.t_addr_prefix == coin.p2sh_t_addr_prefix && coin.segwit {
-        Builder::build_p2sh(&to.hash).to_bytes()
-    } else {
-        return ERR!("Address {} has invalid format", to);
-    };
-    let _utxo_lock = UTXO_LOCK.lock().await;
-    let unspents = try_s!(coin.rpc_client.list_unspent_ordered(&coin.my_address).map_err(|e| ERRL!("{}", e)).compat().await);
-    let (value, fee_policy) = if req.max {
-        (unspents.iter().fold(0, |sum, unspent| sum + unspent.value), FeePolicy::DeductFromOutput(0))
-    } else {
-        (try_s!(sat_from_big_decimal(&req.amount, coin.decimals)), FeePolicy::SendExact)
-    };
-    let outputs = vec![TransactionOutput {
-        value,
-        script_pubkey,
-    }];
-    let fee = match req.fee {
-        Some(WithdrawFee::UtxoFixed { amount }) => Some(ActualTxFee::Fixed(try_s!(sat_from_big_decimal(&amount, coin.decimals)))),
-        Some(WithdrawFee::UtxoPerKbyte { amount }) => Some(ActualTxFee::Dynamic(try_s!(sat_from_big_decimal(&amount, coin.decimals)))),
-        Some(_) => return ERR!("Unsupported input fee type"),
-        None => None,
-    };
-    let (unsigned, data) = try_s!(coin.generate_transaction(unspents, outputs, fee_policy, fee).await);
-    let prev_script = Builder::build_p2pkh(&coin.my_address.hash);
-    let signed = try_s!(sign_tx(unsigned, &coin.key_pair, prev_script, coin.signature_version, coin.fork_id));
-    let fee_details = UtxoFeeDetails {
-        amount: big_decimal_from_sat(data.fee_amount as i64, coin.decimals),
-    };
-    let my_address = try_s!(coin.my_address());
-    let to_address = try_s!(coin.display_address(&to));
-    Ok(TransactionDetails {
-        from: vec![my_address],
-        to: vec![to_address],
-        total_amount: big_decimal_from_sat(data.spent_by_me as i64, coin.decimals),
-        spent_by_me: big_decimal_from_sat(data.spent_by_me as i64, coin.decimals),
-        received_by_me: big_decimal_from_sat(data.received_by_me as i64, coin.decimals),
-        my_balance_change: big_decimal_from_sat(data.received_by_me as i64 - data.spent_by_me as i64, coin.decimals),
-        tx_hash: signed.hash().reversed().to_vec().into(),
-        tx_hex: serialize(&signed).into(),
-        fee_details: Some(fee_details.into()),
-        block_height: 0,
-        coin: coin.ticker.clone(),
-        internal_id: vec![].into(),
-        timestamp: now_ms() / 1000,
-    })
-}
+// We can use a shared UTXO lock for all UTXO coins at 1 time.
+// It's highly likely that we won't experience any issues with it as we won't need to send "a lot" of transactions concurrently.
+lazy_static! {pub static ref UTXO_LOCK: AsyncMutex<()> = AsyncMutex::new(());}
 
-#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
-pub struct UtxoFeeDetails {
-    amount: BigDecimal,
+pub trait UtxoArcGetter {
```

Comment:
Standard AsRef trait looks like good replacement for this: https://doc.rust-lang.org/std/convert/trait.AsRef.html

Reply:
I thought we decide to make a custom trait instead of Deref<UtxoCoinFields>.
What's the best way?

Reply:
`AsRef` and `Deref` are different things 🙂 While `Deref` is recommended to be used only for smart pointers `AsRef` is `reference-to-reference conversion` which looks fine in this case.